apps/SparseMatrix/sparse_matrix.cuh

```diff
@@ -240,6 +241,54 @@ struct SparseMatInfo
                               cudaMemcpyHostToDevice));
     }
 
+    __host__ __device__ T& operator()(const VertexHandle row_v,
+                                      const VertexHandle col_v)
+    {
+        auto     r_ids      = row_v.unpack();
+        uint32_t r_patch_id = r_ids.first;
+        uint16_t r_local_id = r_ids.second;
+
+        auto     c_ids      = col_v.unpack();
+        uint32_t c_patch_id = c_ids.first;
+        uint16_t c_local_id = c_ids.second;
+
+        uint32_t col_index = m_d_patch_ptr_v[c_patch_id] + c_local_id;
+        uint32_t row_index = m_d_patch_ptr_v[r_patch_id] + r_local_id;
+
+        for (uint32_t i = m_d_row_ptr[row_index];
+             i < m_d_row_ptr[row_index + 1];
+             ++i) {
+            if (m_d_col_idx[i] == col_index) {
+                return m_d_val[i];
+            }
+        }
+        assert(1 != 1);
+    }
+
+    __host__ __device__ T& operator()(const VertexHandle row_v,
+                                      const VertexHandle col_v) const
+    {
+        auto     r_ids      = row_v.unpack();
+        uint32_t r_patch_id = r_ids.first;
+        uint16_t r_local_id = r_ids.second;
+
+        auto     c_ids      = col_v.unpack();
+        uint32_t c_patch_id = c_ids.first;
+        uint16_t c_local_id = c_ids.second;
+
+        uint32_t col_index = m_d_patch_ptr_v[c_patch_id] + c_local_id;
+        uint32_t row_index = m_d_patch_ptr_v[r_patch_id] + r_local_id;
+
+        for (uint32_t i = m_d_row_ptr[row_index];
+             i < m_d_row_ptr[row_index + 1];
+             ++i) {
+            if (m_d_col_idx[i] == col_index) {
+                return m_d_val[i];
+            }
+        }
+        assert(1 != 1);
+    }
+
```

Comment:
Rather than duplicating the same code, you may consider having a function that takes `row_v` and `col_v` that returns the index that goes into the `m_d_val`. This function will implement the same logic you have there

Reply:
Do you mean I can change these two functions to three functions where one would take `row_v` and `col_v` that returns the index that goes into the `m_d_val`; and the other two would take `m_d_val` and returns the actual value?

Reply:
Yes, maybe something like that 

```c++
__host__ __device__ uint32_t get_index (const VertexHandle row_v, const VertexHandle col_v) const
{
    //your implementation goes here 
}

__host__ __device__ T& operator()(const VertexHandle row_v, const VertexHandle col_v) 
{
      return m_d_val[get_index (row_v, col_v)];
}

__host__ __device__ T& operator()(const VertexHandle row_v, const VertexHandle col_v) const
{
    return m_d_val[get_index (row_v, col_v)];
}
```

Notice how `get_index` is `const` so you could call it from a const and non-const functions.

Reply:
Yes. I think I have done that. Thanks for the clarification.